### PR TITLE
[openvpn] ccd fixes

### PR DESCRIPTION
**Pull request by vitaliy-sn**
Signed-off-by: Vitaliy Snurnitsin <vitaliy.snurnitsin@flant.com>

## Description

* Fixed feature to assign static addresses to clients.
* In the pushToClientRoutes field, you can now use masks in the formats `10.0.0.0/16` and `10.0.0.0/255.255.0.0`.

## Why do we need it, and what problem does it solve?

The fix will allow the use of static addresses.

## Checklist
- [x] The code is covered by unit tests.
- [ ] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: openvpn
type: feature
summary: Added the ability to use bitmasks in `pushToClientRoutes`. Fixed the feature of assigning static addresses for VPN clients.
```

### CODE REVIEW

File: modules/500-openvpn/templates/openvpn/configmap.yaml [hunk 1]
```diff
@@ -88,12 +83,30 @@ data:
     push "dhcp-option DOMAIN {{ $searchDomain }}"
       {{- end }}
     {{- end }}
+  sync_ccd_udp.sh: |-
```
Comment (apolovov) [inline modules/500-openvpn/templates/openvpn/configmap.yaml:86]: don't we need sync_ccd_tcp?

**Comment by apolovov** [comment]
Please, fill the `Why do we need it, and what problem does it solve?` section.

File: modules/500-openvpn/templates/openvpn/configmap.yaml (hunk 1 above)
Comment (vitaliy-sn) [reply modules/500-openvpn/templates/openvpn/configmap.yaml]: Made a common script for tcp and udp

File: modules/500-openvpn/templates/openvpn/configmap.yaml [hunk 2]
```diff
@@ -88,12 +83,45 @@ data:
...
+    tcp_enabled="$1"
+    udp_enabled="$2"
+    tcp_network="$(echo $3 | rev | sed 's/\./ /' | rev | awk '{print $1}')"
+    udp_network="$(echo $4 | rev | sed 's/\./ /' | rev | awk '{print $1}')"
+
+
+    while true; do
+      if [ "$tcp_enabled" == "true" ]; then
+        if [ -d /etc/openvpn/ccd ]; then
+          for f in $(ls -1 /etc/openvpn/ccd); do
```
Comment (apolovov) [inline modules/500-openvpn/templates/openvpn/configmap.yaml:99]:
```suggestion
          for f in /etc/openvpn/ccd/*; do
```

**Review by apolovov — CHANGES_REQUESTED** [review]

File: modules/500-openvpn/docs/CONFIGURATION_RU.md [hunk 3]
```diff
@@ -9,6 +9,8 @@ data:
   openvpnEnabled: "true"
 ```
 
+**Внимание!** В админке при назначении статического адреса нужно выбирать адрес из диапазона tunnelNetwork. Подсети для TCP и UDP сервера должны быть одного размера. Статические адреса в CCD (Client-specific configuration) для UDP сервера преобразутся из CCD для TCP подменой подсети.
+
```
Comment (z9r5) [inline modules/500-openvpn/docs/CONFIGURATION_RU.md:13]:
```suggestion
**Внимание!** В панели администратора при назначении статического адреса нужно выбирать адрес из диапазона tunnelNetwork. Подсети для TCP и UDP-сервера должны быть одного размера. Статические адреса в CCD (Client-specific configuration) для UDP-сервера преобразутся из CCD для TCP подменой подсети.

```

**Review by z9r5 — APPROVED** [review]

File: modules/500-openvpn/templates/openvpn/statefullset.yaml [hunk 4]
```diff
@@ -147,22 +157,42 @@ spec:
...
             readOnly: true
-          - name: entrypoint
+          - name: scripts
             mountPath: /entrypoint.sh
             subPath: entrypoint.sh
             readOnly: true
         resources:
           requests:
-{{- include "helm_lib_module_ephemeral_storage_logs_with_extra" 10 | indent 12 }}
+            {{- include "helm_lib_module_ephemeral_storage_logs_with_extra" 10 | indent 12 }}
```
[the remark this answers is not in the document]
Comment (vitaliy-sn) [reply modules/500-openvpn/templates/openvpn/statefullset.yaml]: fixed

File: modules/500-openvpn/templates/openvpn/statefullset.yaml [hunk 5]
```diff
@@ -118,13 +128,13 @@ spec:
...
```
[the remark this answers is not in the document]
Comment (vitaliy-sn) [reply modules/500-openvpn/templates/openvpn/statefullset.yaml]: fixed

File: modules/500-openvpn/templates/openvpn/statefullset.yaml [hunk 6]
```diff
@@ -177,6 +207,7 @@ spec:
             --listen.host="127.0.0.1"
             --listen.port="8000"
             --role="master"
+            --ovpn.network="{{ include "get_network_with_bitmask" (list . .Values.openvpn.tunnelNetwork) }}"
```
[the remark this answers is not in the document]
Comment (vitaliy-sn) [reply modules/500-openvpn/templates/openvpn/statefullset.yaml]: After the last discussion, we decided to always use tunnelNetwork and add information about this to the documentation

File: modules/500-openvpn/templates/openvpn/statefullset.yaml [hunk 7]
```diff
@@ -283,6 +314,10 @@ spec:
         emptyDir:
           medium: Memory
           sizeLimit: 128Mi
+      - name: ccd-udp
```
[the remark this answers is not in the document]
Comment (vitaliy-sn) [reply modules/500-openvpn/templates/openvpn/statefullset.yaml]: After the last changes, the ccd-tcp (original files) and ccd-udp directories remained.

File: modules/500-openvpn/templates/openvpn/statefullset.yaml [hunk 8]
```diff
@@ -147,22 +157,42 @@ spec:
...
+            mountPath: /etc/openvpn/ccd
+          - name: ccd-udp
+            mountPath: /etc/openvpn/ccd-udp
+          - name: scripts
+            mountPath: /sync_ccd_udp.sh
+            subPath: sync_ccd_udp.sh
+            readOnly: true
+        resources:
+          requests:
+            {{- include "helm_lib_module_ephemeral_storage_logs_with_extra" 10 | indent 12 }}
```
[the remark this answers is not in the document]
Comment (vitaliy-sn) [reply modules/500-openvpn/templates/openvpn/statefullset.yaml]: fixed

File: modules/500-openvpn/docs/CONFIGURATION.md [hunk 9]
```diff
@@ -9,6 +9,8 @@ data:
   openvpnEnabled: "true"
 ```
 
+**Caution!** The admin panel always uses tunnelNetwork. Static addresses must be issued from it. If UDP openvpn server is enabled, tunnelNetwork and udpTunnelNetwork must be the same size.
```
Comment (apolovov) [inline modules/500-openvpn/docs/CONFIGURATION.md:12]:
```suggestion
**Caution!** The admin panel always uses tunnelNetwork (not udpTunnelNetwork one). Static addresses must be issued from it. If UDP openvpn server is enabled,  tunnelNetwork and udpTunnelNetwork must be the same size.
```

File: modules/500-openvpn/docs/CONFIGURATION_RU.md [hunk 10]
```diff
@@ -9,6 +9,8 @@ data:
   openvpnEnabled: "true"
 ```
 
+**Внимание!** В панели администратора всегда используется tunnelNetwork, статические адреса необходимо выдавать из неё. Если включен UDP openvpn сервер, то tunnelNetwork и udpTunnelNetwork должны быть одного размера.
```
Comment (apolovov) [inline modules/500-openvpn/docs/CONFIGURATION_RU.md:12]:
```suggestion
**Внимание!** В панели администратора всегда используется tunnelNetwork (но не udpTunnelNetwork), статические адреса необходимо выдавать из неё. Если включен UDP openvpn сервер, то tunnelNetwork и udpTunnelNetwork должны быть одного размера.
```

**Review by apolovov — APPROVED** [review]

**Pull request merged**